master/internal/command/command.go

```diff
@@ -146,357 +150,311 @@ func remakeCommandsByType(
-		err := task.DefaultService.Signal(c.allocationID, task.KillAllocation, "user requested kill")
-		if err != nil {
-			ctx.Log().WithError(err).Warn("failed to kill allocation")
-		}
-		ctx.Respond(&apiv1.KillCommandResponse{Command: c.toCommand(ctx)})
+// snapshotAndPersist turns the command into a CommandSnapshot (see command/models.go) & persists that to the db.
+func (c *command) snapshotAndPersist(ctx context.Context) error {
+	snapshot := &CommandSnapshot{
+		TaskID:             c.taskID,
+		RegisteredTime:     c.registeredTime,
+		AllocationID:       c.allocationID,
+		GenericCommandSpec: c.GenericCommandSpec,
+	}
+	_, err := db.Bun().NewInsert().Model(snapshot).
+		On("CONFLICT (task_id) DO UPDATE").
+		Exec(ctx)
+	return err
+}
 
-	case *apiv1.SetCommandPriorityRequest:
-		err := c.setPriority(int(msg.Priority), true)
-		if err != nil {
-			ctx.Respond(err)
-			return nil
-		}
-		ctx.Respond(&apiv1.SetCommandPriorityResponse{Command: c.toCommand(ctx)})
+// onExit runs when an command's allocation exits. It marks the command task as complete, and unregisters where needed.
+// onExit locks ahead of gc -> unregisterCommand.
+func (c *command) onExit(ctx context.Context, ae *task.AllocationExited) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
 
-	case *shellv1.Shell:
-		ctx.Respond(c.toShell(ctx))
+	c.exitStatus = ae
 
-	case *apiv1.GetShellRequest:
-		ctx.Respond(&apiv1.GetShellResponse{
-			Shell:  c.toShell(ctx),
-			Config: protoutils.ToStruct(c.Config),
-		})
+	if err := c.db.CompleteTask(c.taskID, time.Now().UTC()); err != nil {
+		c.syslog.WithError(err).Error("marking task complete")
+	}
 
-	case *apiv1.KillShellRequest:
-		err := task.DefaultService.Signal(c.allocationID, task.KillAllocation, "user requested kill")
-		if err != nil {
-			ctx.Log().WithError(err).Warn("failed to kill allocation")
-		}
-		ctx.Respond(&apiv1.KillShellResponse{Shell: c.toShell(ctx)})
+	if err := user.DeleteSessionByToken(ctx, c.GenericCommandSpec.Base.UserSessionToken); err != nil {
+		c.syslog.WithError(err).Errorf("failure to delete user session for task: %v", c.taskID)
+	}
 
-	case *apiv1.SetShellPriorityRequest:
-		err := c.setPriority(int(msg.Priority), true)
-		if err != nil {
-			ctx.Respond(err)
-			return nil
+	go func() {
+		time.Sleep(terminatedDuration)
+		c.gc(ctx)
+	}()
+}
+
+// gc garbage collects the exited command.
+func (c *command) gc(ctx context.Context) {
+	if err := tasklist.GroupPriorityChangeRegistry.Delete(c.jobID); err != nil {
+		c.syslog.WithError(err).Error("deleting command from GroupPriorityChangeRegistry")
+	}
+	if c.exitStatus == nil {
+		if err := c.db.CompleteTask(c.taskID, time.Now().UTC()); err != nil {
+			c.syslog.WithError(err).Error("marking task complete")
 		}
-		ctx.Respond(&apiv1.SetShellPriorityResponse{Shell: c.toShell(ctx)})
+	}
+	// TODO CAROLINA: running these unregister calls as goroutines for now.
+	go job.DefaultService.UnregisterJob(c.jobID)
+	go DefaultCmdService.unregisterCommand(c.taskID)
```

Comment:
i'd have a different strategy for stuff like this, maybe we (you me erik) can talk offline now that I've looked?